course3/homework-linkedlist/linkedlist.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+#include <iostream>
+#include <memory>
+
+namespace mau {
+
+// Reverse iterator. http://stackoverflow.com/questions/8542591/c11-reverse-range-based-for-loop
+template <typename T>
+struct reversion_wrapper { T& iterable; };
+
+template <typename T>
+auto begin (reversion_wrapper<T> w) { return rbegin(w.iterable); }
+
+template <typename T>
+auto end (reversion_wrapper<T> w) { return rend(w.iterable); }
+
+template <typename T>
+reversion_wrapper<T> reverse (T&& iterable) { return { iterable }; }
+
+template<typename T>
+class list {
+public:
+    typedef ptrdiff_t                          difference_type;
+    typedef std::forward_iterator_tag          iterator_category;
+    typedef T                                  value_type;
+    typedef T*                                 pointer;
+    typedef T&                                 reference;
+    class Node {
+    public:
+        Node(const T& val, std::unique_ptr<Node>& next) : mData(val), mNext(std::move(next)) {}
+        T mData;
+        std::unique_ptr<Node> mNext;
+    };
+
+    class iterator {
+    public:
+
+        iterator(Node* node) : mNode(node) {}
+        bool operator==(const iterator& it) { return mNode == it.mNode; }
+        bool operator!=(const iterator& it) { return mNode != it.mNode; }
+        iterator& operator++() { 
+            mNode = mNode->mNext.get();
+            return *this;
+        }
+        T& operator*() {
+            return mNode->mData;
+        }
+
+        Node* mNode = nullptr;
+
+    };
+
+    class const_iterator {
+    public:
+        const_iterator(const iterator& it) : mNode(it.mNode) {}
+        const_iterator(Node* node) : mNode(node) {}
+        bool operator==(const const_iterator& it) { return mNode == it.mNode; }
+        bool operator!=(const const_iterator& it) { return mNode != it.mNode; }
+        const_iterator& operator++() { 
+            mNode = mNode->mNext.get();
+            return *this;
+        }
+        T& operator*() {
+            return mNode->mData;
+        }
+
+        Node* mNode = nullptr;
+    };
+
+    list() : mSize() {}
+
+    list(std::initializer_list<T> init) : mSize(0) {
```

Comment:
This is fine, but this will only be selected for initializer_lists of objects of exactly the type ```T```.
Think of the following use-cases:
```
mau::list<std::string> ls = {"abc", "def"}; // T=std::string, but initialized with std::initializer_list<const char*>
mau::list<double> ld = {1, 3, 5}; // T=double, but initialized with std::initializer_list<double>
```